crates/ursa-rpc-service/src/api.rs

```diff
@@ -79,68 +88,228 @@ pub trait NetworkInterface: Sync + Send + 'static {
     /// Put a car file and start providing to the network
     async fn put_car<R: AsyncRead + Send + Unpin>(&self, file: Car<R>) -> Result<Vec<Cid>>;
 
-    // Put a file using a local path
+    /// Put a file using a local path
     async fn put_file(&self, path: String) -> Result<Vec<Cid>>;
 
-    // get peers from the network
+    /// Get peers from the network
     async fn get_peers(&self) -> Result<HashSet<PeerId>>;
 
-    // get the addrresses that p2p node is listening on
+    /// Get the addresses that p2p node is listening on
     async fn get_listener_addresses(&self) -> Result<Vec<Multiaddr>>;
 }
-#[derive(Clone)]
+
+type PendingRequests = Arc<RwLock<HashMap<Cid, Vec<Sender<Result<()>>>>>>;
+
 pub struct NodeNetworkInterface<S>
 where
     S: Blockstore + Store + Send + Sync + 'static,
 {
     pub store: Arc<UrsaStore<S>>,
     pub network_send: Sender<NetworkCommand>,
     pub provider_send: Sender<ProviderCommand>,
+    pending_requests: PendingRequests,
+    client: Arc<Client>,
+    origin_config: OriginConfig,
 }
 
-#[async_trait]
-impl<S> NetworkInterface for NodeNetworkInterface<S>
+impl<S> NodeNetworkInterface<S>
 where
     S: Blockstore + Store + Send + Sync + 'static,
 {
-    async fn get(&self, cid: Cid) -> Result<Option<Vec<u8>>> {
+    pub fn new(
+        store: Arc<UrsaStore<S>>,
+        network_send: Sender<NetworkCommand>,
+        provider_send: Sender<ProviderCommand>,
+        origin_config: OriginConfig,
+    ) -> Self {
+        Self {
+            store,
+            network_send,
+            provider_send,
+            origin_config,
+            pending_requests: Arc::new(RwLock::new(HashMap::new())),
+            client: Arc::new(Client::new()),
+        }
+    }
+
+    /// Check blockstore for content. On cache miss, attempt to fetch and start providing it
+    /// Returns Ok(true) if the block was found in the blockstore, and Ok(false) if the content was fetched
+    async fn check_blockstore(&self, cid: Cid) -> Result<bool> {
```

Comment:
should we `check_blocstore` have one job instead of a nested `get_network`. Like what if I just want to check if a block exists

Reply:
Maybe a better name for this fn would be better, like `check_or_fetch` ?

Reply:
Both `get` and `get_data` had the same code that is now check_blockstore, I put it in a function so as to DRY

Reply:
I see that now, hard to see when block of code are just moved around arbitrarily

Reply:
nothing was moved around it was put into an internal implementation because not used externally as recommended by https://github.com/fleek-network/ursa/pull/278#discussion_r1068751470

Reply:
You mentioned it was inserted in between so it was moved

Reply:
Unresolved this, the original comment wasn't addressed

Reply:
I'll update the method with a better name, to clearer communicate what it's purpose is

> what if I just want to check if a block exists

Just using `store.blockstore().has(cid)` should suffice for that usecase

Reply:
Renamed `check_blockstore` to `sync_content` here:
https://github.com/fleek-network/ursa/pull/278/commits/ebb750854c81e7ec038c6e740b0cc8db0d4a3085